### PR TITLE
Add stats for `query` and `cache` requests

### PR DESCRIPTION
**Pull request by GKosheev**
Issue #64  

I implemented ```get``` method to get the size of cached calls: 

```javascript
get numberOfCachedCalls() {
    return this._cache.size
}
```

### CODE REVIEW

**Comment by szmarczak** [comment]
This is actually incorrect. It doesn't return the total number of cached requests, but the current cache size. I'd implement it like this:

```js
this.stats = {
	cache: 0,
	query: 0,
};
```

and increment the counters respectively in https://github.com/szmarczak/cacheable-lookup/blob/5b08a3610a5ec1328da583a6327111428ff85c42/source/index.js#L213

You also need to provide some tests.

**Comment by GKosheev** [comment]
In the last commit, I erased created method ```numberOfCachedCalls``` and added counters to ```query``` method.

Also, I implemented unit test: 
```javascript
test('cache and query stats', async t => {
	const cacheable = new CacheableLookup();
	t.is(cacheable.stats.query, 0);
	t.is(cacheable.stats.cache, 0);

	await cacheable.lookupAsync('1dot1dot1dot1.cloudflare-dns.com');
	t.is(cacheable.stats.query, 1);
	t.is(cacheable.stats.cache, 1);

	await cacheable.lookupAsync('1dot1dot1dot1.cloudflare-dns.com');
	t.is(cacheable.stats.query, 2);
	t.is(cacheable.stats.cache, 1);

	await cacheable.lookupAsync('google.com');
	t.is(cacheable.stats.query, 3);
	t.is(cacheable.stats.cache, 2);
});

```

**Comment by szmarczak** [comment]
@GKosheev No need to duplicate code in the comments :) There's a `review` feature and you can add comments there if necessary.

File: source/index.js [hunk 1]
```diff
@@ -211,19 +215,19 @@ class CacheableLookup {
 	}
 
 	async query(hostname) {
+		this.stats.query++;
```
Comment (szmarczak) [inline source/index.js:218]: This is incorrect. `query` should be only incremented when `source` is set to `query`.

File: source/index.js [hunk 2]
```diff
@@ -211,19 +215,19 @@ class CacheableLookup {
...
 			const pending = this._pending[hostname];
-
 			if (pending) {
 				cached = await pending;
 			} else {
 				source = 'query';
 				const newPromise = this.queryAndCache(hostname);
 				this._pending[hostname] = newPromise;
-
+				this.stats.cache++;
```
Comment (szmarczak) [inline source/index.js:230]: source is set to `query` but you increment `cache` instead.

Comment (szmarczak) [reply source/index.js:230]:
```suggestion
				this.stats.query++;
```

File: source/index.js (hunk 1 above)
Comment (szmarczak) [reply source/index.js:218]:
```suggestion
```

File: source/index.js [hunk 3]
```diff
@@ -211,19 +215,19 @@ class CacheableLookup {
...
 	async query(hostname) {
+		this.stats.query++;
 		let source = 'cache';
 		let cached = await this._cache.get(hostname);
 
 		if (!cached) {
 			const pending = this._pending[hostname];
-
 			if (pending) {
 				cached = await pending;
```
Comment (szmarczak) [inline source/index.js:225]:
```suggestion
				this.stats.cache++;
				cached = await pending;
```

File: source/index.js [hunk 4]
```diff
@@ -211,19 +215,19 @@ class CacheableLookup {
 	}
 
 	async query(hostname) {
+		this.stats.query++;
 		let source = 'cache';
 		let cached = await this._cache.get(hostname);
 
 		if (!cached) {
```
Comment (szmarczak) [inline source/index.js:222]:
```suggestion
		if (cached) {
			this.stats.cache++;
		} else {
```

File: tests/test.js [hunk 5]
```diff
@@ -1067,3 +1067,21 @@ test.cb.failing('throws original lookup error if not recognized', t => {
 		t.end();
 	});
 });
+
+test('cache and query stats', async t => {
+	const cacheable = new CacheableLookup();
```
Comment (szmarczak) [inline tests/test.js:1072]: Can you use the mocked resolver like in the tests above?

**Review by szmarczak — CHANGES_REQUESTED** [review]

File: source/index.js (hunk 4 above)
Comment (GKosheev) [reply source/index.js:222]: I don't think I understand this particular suggestion quite well on top of others. 
Do you want me to get rid of ```!``` and make a different ``` if``` statement?

File: tests/test.js (hunk 5 above)
Comment (GKosheev) [reply tests/test.js:1072]: I'm trying to understand how to use ```resolver```. 
Would it be enough if I just add it to the existing test like that?
```javascript
const cacheable = new CacheableLookup({resolver});
```

Comment (szmarczak) [reply tests/test.js:1072]: Here's an example: https://github.com/GKosheev/cacheable-lookup/blob/ca90ec3857e219af14c7ec7c2b9580c5de5f2601/tests/test.js#L408-L430

**Comment by GKosheev** [comment]
I'm not sure if I did the unit test as you wanted me to, I tried to follow the example you sent me with the ```verify``` method.  
But I am always ready to make corrections.

File: tests/test.js [hunk 6]
```diff
@@ -1067,3 +1067,33 @@ test.cb.failing('throws original lookup error if not recognized', t => {
 		t.end();
 	});
 });
+
+test('cache and query stats', async t => {
+	const cacheable = new CacheableLookup({resolver});
+
+	let entries = await cacheable.lookupAsync('1dot1dot1dot1.cloudflare-dns.com', {all: true, family: 4});
```
Comment (szmarczak) [inline tests/test.js:1074]: Please use a hostname present in the mocked resolver

File: tests/test.js [hunk 7]
```diff
@@ -1067,3 +1067,33 @@ test.cb.failing('throws original lookup error if not recognized', t => {
...
+	verify(t, entries, [
+		{family: 4, source: 'query'}
```
Comment (szmarczak) [inline tests/test.js:1076]: Missing `address`

**Review by szmarczak — CHANGES_REQUESTED** [review]

**Comment by GKosheev** [comment]
In the last commit, I added the missing ```address``` and changed the hostname to ```temporary```.

**Comment by GKosheev** [comment]
@szmarczak, would that work?

**Comment by szmarczak** [comment]
Will take a look now, sorry for delay. Also, no need to post a comment of your commit changes :)

**Review by szmarczak — APPROVED** [review]

**Comment by szmarczak** [comment]
Looks good! Thanks for contributing! :heart:

**Comment by wichopy** [comment]
@szmarczak @GKosheev I dont see an api in the docs to access the stats, do I just do cacheable.stats ??

**Pull request merged**